### DISCUSSION TITLE
AMR Annotation and Pandora's Analysis Report

### DISCUSSION THREAD

**Discussion by SARDONYX-sard** [Monitor221hz/Pandora-Behaviour-Engine-Plus#178] [Ideas]
Here we would like to consider how to apply AMR annotation data in Pandra.
If that is possible, Patch will no longer be necessary.

## Related Issues 
- [TK Dodge SE/Standalone Requires Patch (Pandora Will Only Support Motion Data via AMR)](#28)

## CrouchSlide_slide$0

![CrouchSlide_slide$0](https://github.com/Monitor144hz/Pandora-Behaviour-Engine-Plus/assets/68905624/6b7f2709-65a0-4361-9bd1-b569abf155d3)

## CrouchSlide_xml_diff(Generated by [hkxconv](https://github.com/ret2end/hkxconv))

          ← CrouchSlide Pandra patch                        /                 CrouchSlide Normal →
![CrouchSlide_xml_diff](https://github.com/Monitor144hz/Pandora-Behaviour-Engine-Plus/assets/68905624/91c433ae-dc03-4740-b699-31f879515ca4)

## animationdatasinglefile_diff

          ← Files generated by Pandra                        /                 Files generated by Nemesis →
![CrouchSlide_animationdatasinglefile_diff](https://github.com/Monitor144hz/Pandora-Behaviour-Engine-Plus/assets/68905624/a79767a2-2f78-48b2-be9a-a12d83c844f6)

## Result

If we can do either of the following from this image with this engine, we might be able to avoid making your own Patch.

- Modify the XML as in the AMR annotation patch.
- animationdatasinglefile to be the same as the file generated by Nemesis

Unfortunately, I don't know if that is possible as I can write some code but have little knowledge of C# and Havok Animation.

**Comment by lobotomy-x** [comment]
several ways to go about this, it really just depends on how monitor wants to incorporate it and whether we can programmatically determine the right timings for annotations at a universal level. I supposed you could expose some options in the GUI but in any case I think we could work from [Hkx/PackFileValidator.cs](https://github.com/Monitor144hz/Pandora-Behaviour-Engine-Plus/blob/main/PandoraPlus/MVVM/Model/Patch/Patchers/Skyrim/Hkx/PackFileValidator.cs)  or  somewhere else after decoding to xml, search for any markers of motion and confirm that we don't have amr annotations already in there and then I think if you can determine when to add them you could just add them to the xml by calling the insertelementchange class. I believe that would be equivalent to adding annotations with hkanno and would be preferable since we're already in the files but alternatively you can bundle hkanno64 and call it from cs like so 

```
          Trace.WriteLine("\n Working on : " + str + "\n");
          for (int index = 0; index < strArray.Length; ++index)
          {
            if (!strArray[index].ToLower().Contains("animmotion") && !strArray[index].ToLower().Contains("#"))
              Trace.WriteLine(strArray[index]);
           }

          if (File.Exists("anno.txt"))
            File.Delete("anno.txt");
          File.WriteAllLines("anno.txt", strArray);
          Process.Start("cmd.exe", "/C" + ("hkanno64.exe update -i anno.txt " + str)).WaitForExit();
        }
```

obviously this doesn't contain the logic for determining which animations to patch or how to populate the annotation array but its a functional jumping off point if you want to try building it yourself. I am interested in the idea of programmatically determining movement direction to automatically assign keytrace conditions for OAR so it would be cool if we could figure that out. Otherwise imo the best solution is probably to stick with some degree of manual preparation or bundling patches since new mods should be using amr to begin with. If you want to mess around with it on your own and don't want to use CS you can easily call that command line from python

**Reply by SARDONYX-sard** [reply]
Ah, I hadn't thought about using hkanno64.exe.
Surely we could write our own scripts and batch process them.

I'll have to learn about hkanno64.exe as I've never used it.

**Reply by lobotomy-x** [reply]
hkanno is incredibly easy to use, it has exactly 2 functions. 

`./hkanno64.exe dump -o anno.txt Animation.hkx` and `./hkanno64.exe update -i anno.txt Animation.hkx`


here is a bulk dumper. save this as a .bat wherever you like. it can even be run on your entire active mods by launching from mo2 if you dont mind waiting 30 minutes

`@ECHO OFF
setlocal enabledelayedexpansion
cd /d %~dp0
hkanno = "%~dps0\hkanno64.exe"
Rem set your actual hard path here


for /r %%f in (*.hkx) do (
	
  SET oDir=%%~DPf
  SET oFile=%%~f
  set /p val=<%%f
  echo "fullname: %%f"
  echo "name: %%~nf
  echo "contents: !val!"
  
!hkanno! dump -o %%~dpnf.txt %%f
)

`

if you just want to be able to drag one file onto a bat for processing

`﻿hkanno64.exe dump "%~1"` 

updates follow identical pattern as dump so just look at those for guidance

You will need [havok tools in the same folder or linked in env vars (same folder is easiest)](https://www.softpedia.com/get/Programming/Other-Programming-Files/Havok-Content-Tools.shtml#download)

again idk if we actually need it since pandora already has us poking around in the behaviors and could potentially handle it that way. Well actually now I think about it those are just the behavior graphs and it currently doesnt touch the animations that need annotating but it probably could? OR maybe it does already since I think we convert to 64 bit, not sure. I'd say just use hkanno until monitor weighs in

**Reply by SARDONYX-sard** [reply]
Thanks for the advice.

Looking at the source of hkanno64.exe, I see that this command is editing annotationTracks in the Havok SDK.
It's named "64" but the exe is a 32bit binary because the SDK is for 32bit. (I checked with `dumpbin /headers . \hkanno64.exe`)
![2024-04-05 (1)](https://github.com/Monitor144hz/Pandora-Behaviour-Engine-Plus/assets/68905624/4f7b84df-0beb-456a-b3c9-f7a7597beec9)

 I tried it with Smooth Bow Animation. This may indeed be useful.
![2024-04-05 (2)](https://github.com/Monitor144hz/Pandora-Behaviour-Engine-Plus/assets/68905624/cb5b3000-8109-42bc-8037-8e6688dc38f9)

```powershell
# Get the directory of the script
$scriptDirectory = Split-Path -Parent $MyInvocation.MyCommand.Path

# Set the path to hkanno64.exe
$hkanno = Join-Path $scriptDirectory "hkanno64.exe"

# Get all .hkx files recursively in the script directory
$files = Get-ChildItem -Path $scriptDirectory -Filter "*.hkx" -Recurse

foreach ($file in $files) {
    # Get the content of the file
    $content = Get-Content -Path $file.FullName

    Write-Output "fullname: $($file.FullName)"
    Write-Output "name: $($file.BaseName)"
    Write-Output "contents: $content"

    # Execute hkanno64.exe with dump command
    & $hkanno dump -o "$($file.DirectoryName)\$($file.BaseName).txt" $file.FullName
}
```

I learned a lot. Thank you.

**Comment by lobotomy-x** [comment]
yes its 32bit application made to run 64bit havok (which you must provide yourself and is not bundled with it, I did give you a link to it though) and produce 64 bit animations. there's no actual reason for hkanno to be 64 bit since it should never use much memory. nemesis similarly is 32 bit but produces 64 bit hkx files for skyrim, however that would be an example of an app that benefits from 64 bit since it crashes so much. It looks like you pretty much just translated the script I provided to powershell with a few changes. Were you running into issues with what I gave or is ps just preference?

**Reply by SARDONYX-sard** [reply]
Certainly the same was true of Nemesis.

Yes, I used PowerShell because I can hardly read bat.

**Comment by SARDONYX-sard** [comment]
As a result of creating my own replacement software, animationsingledatafile.txt is mostly an add patch, and it can be run simply by adding it appropriately. However, there is probably an upper limit due to ID depletion. (i32::MAX?)

My question has been answered, so I will close this thread.
